### PR TITLE
[Filesystem] Fix makePathRelative when the original path does not end with a slash

### LINKED ISSUE #25023: [Filesystem] Several issues with Filesystem::makePathRelative
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | master

Looking at the code of Filesystem::makePathRelative method I have found several issues:
- Stripping of drive letter for arguments is done regardless of host OS but directory separator normalization is done only for Windows hosts.
- There is no drive check for Windows paths, hence for the following case result will be invalid: `(new Filesystem())->makePathRelative('C:/aa/bb/cc', 'D:/aa/dd/..')` produces `'bb/cc/'`
- `(new Filesystem())->makePathRelative('/aa/bb/cc', '/aa')` produces `'bb/cc/'` which is wrong since `'/aa/bb/cc'` can be the file path, the presence of trailing slash in the result should be based on presence of trailing slash in the `$endPath` argument.

### PR DESCRIPTION
**Pull request by the-csaba**
| Q             | A
| ------------- | ---
| Branch?       |  4.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #25023
| License       | MIT
| Doc PR        | N/A

Maintain the path format after making relative. 

History: 
PR #4842 added a slash to the function output. According to the original explanation, "[...]to ensure all returned paths end with /."

Reason for change:
There is no need to add a slash, not even for operation upon directories. Maintaining the input format should enough for all file manipulation (symlink, copy, unset, etc.). There is a slight chance to introduce a break, but only if the path further altered after it changed to the relative. One possible mitigation would be to checking the input; it is a file or directory. But, the function in the current form does not access the filesystem. Adding an extra check will significantly slow it down and produce two different results when the `$endPath` exists, a directory, and the input does not have a trailing slash.

Test altered to reflect the changes.

### CODE REVIEW

File: src/Symfony/Component/Filesystem/Tests/FilesystemTest.php [hunk 1]
```diff
@@ -1114,32 +1114,32 @@ public function providePathsForMakePathRelative()
             ['/aa/bb/cc', '/aa/bb/cc/dd/', '../'],
             ['/aa/bb/cc/', '/aa/bb/cc/dd', '../'],
             ['/aa/bb/cc/', '/aa/bb/cc/dd/', '../'],
-            ['/aa/bb/cc', '/aa', 'bb/cc/'],
```
Comment (jderusse) [inline src/Symfony/Component/Filesystem/Tests/FilesystemTest.php:1117]: Please don't delete previous tests. 
Only add new cases

Comment (the-csaba) [reply src/Symfony/Component/Filesystem/Tests/FilesystemTest.php:1117]: It changed not deleted. Please see my explanation in the description. Without changing the expectation, we can't fix the method.

Comment (jderusse) [reply src/Symfony/Component/Filesystem/Tests/FilesystemTest.php:1117]: indeed, I misread the description.

**Review by Skydrol — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
> There is a slight chance to introduce a break, but only if the path further altered after it changed to the relative. One possible mitigation would be to checking the input; it is a file or directory

I think the check is worth it. is_dir is cached, so it won't slow down things much.

**Comment by the-csaba** [comment]
> I think the check is worth it. is_dir is cached, so it won't slow down things much.

You mean we will be checking the directory and adding a slash regardless of the input string?

I am assuming the method will still  have output when a path not exists, then it can produce different output on subsequent run or different state of the filesystem:

```php
$filesystem =new Symfony\Component\Filesystem\Filesystem();
$relative = $filesystem->makePathRelative('/path/to/project/resources', '/path/to/project');

if ($filesystem->exists('/path/to/project/resources')) {
    mkdir('/path/to/project/resources' );
}
echo $relative . PHP_EOL;

// -> '/path/to/project/resources'  if dir not exist
// -> '/path/to/project/resources/' if dir exist (or second run)
```
I believe it will create more problems than we currently have. 
So I think changing the output is a better way forward.

**Comment by nicolas-grekas** [comment]
I'm closing here because I think this can create troubles in existing codebases.
I agree it could have been a better, but the current behavior is too old to break.
Also, there are ways around (eg using the method on `dirname($file)`)
Thanks for submitting.

**Comment by the-csaba** [comment]
I understand the reasons.

Given the `$relativePath` is the output of the `makePathRelative()` method, and `$endPath` is the first argument, the following snippet removes the trailing slash if not present in the input:

```php
if (!str_ends_with($endPath, '/')) {
    $relativePath =  rtrim($relativePath, '/');
}
```

**Comment by mvorisek** [comment]
This needs to be definitely fixed, when a non-dir path is given (path not ending with `/`), the result must never have `/` appended. I agree this can imply BC break, bug the current impl. is simply wrong. Please reopen this PR or let me know if I or the original author should submit it newly.

**Comment by strarsis** [comment]
+1, I use this function to make a path to a stylesheet file (e.g. `/srv/www/[...]/folder/app.css` relative to another folder, 
currently `makePathRelative` returns the path with a slash at the end (`folder/app.css/`).

**Comment by mvorisek** [comment]
Sensible fix like https://github.com/atk4/ui/blob/8f7a986c629a0adf3342c6d285bbef613fe52095/src/App.php#L655-L658 works /w current/buggy Symfony and /w this PR/fixed Symfony so fixing it should be fine as long as the developer was aware of this issue and counted with the possibility to be fixed in the future.

For me, it is clearly a bug in Symfony and it should be fixed sooner or later.

Can you please reconsider this PR, if not for 4.4, can it be integrated into 6.x?

**Comment by mvorisek** [comment]
@nicolas-grekas Sensible fix like https://github.com/atk4/ui/blob/8f7a986c629a0adf3342c6d285bbef613fe52095/src/App.php#L655-L658 works /w current/buggy Symfony and /w this PR/fixed Symfony so fixing it should be fine as long as the developer was aware of this issue and counted with the possibility to be fixed in the future.

For me, it is clearly a bug in Symfony and it should be fixed sooner or later.

Can you please reconsider this PR, if not for 4.4, can it be integrated into 6.x?

**Pull request closed without merging**